### DISCUSSION TITLE
How to deploy DAGs from git repository to airflow

### DISCUSSION THREAD

**Discussion by vladaman** [apache/airflow#29229] [Q&A]
How do I pull DAGs from our git repository to airflow? We have installed airflow with docker-compose, setup dag/ folder. When we remotely create dag sample files it works. How do you integrate this with your CI/CD pipeline? Do you push this to airflow via scp or you let airflow to pull artifacts from s3 or similar? Is it good idea to let airflow do git pull on our repo? What would be your recommended solution? Also we need to install dependencies for python scripts.

**Comment by potiuk** [comment]
The recommended solution is using a git-sync container to do the job of syncing DAGs from git.  How to do it in docker-compose is up to you. As you can clearly see in our docker-compose quick-start example , you are mostly on your own to build your own docker-compose to handle any kind of productionise setup. This is simply property od docker-compose that you as deployment manager have to craft and define the docker-compose on your own. And communiy help is at most giving you directions like this. 

The recommended solution where community provides much more.of a configurable and working approach is K8S and helm chart we publish (you can find it via  https://airflow.apache.org/docs/helm-chart/stable/index.html )- you can take a look there, git sync is an option that you can just enable and it runs git-sync sidecars to perform syncing with the git repo of yours. You can also use the helm chart as an inspiration for your own docker-compose crafting. 

But i would say that if you really want to have serious production deployment with lots of things handled for you - go k8s and helm, it will save you multiple hours of crafting and handling your own solution with docker-compose (for example it nicely handles ugrades via init containers and has built-in pgbouncer configure if you use postgres and much, much more). All those things you will have to do  on your own with docker-compose,  figure your own ways of performing upgrades that suit your workflows, choose some persistency approach for logs possibly etc. etc. All this is not needed with Helm. And you are responsible for figuring out how to do it when you are choosing docker-compose. Up to you which path you take.

When it comes to dependencies you need to add (as also explained in the compose/helm docs) you need to build your own image - either by extending or customisig the base Airflowa image - details here https://airflow.apache.org/docs/docker-stack/build.html - this is the only recommended way, some people try to dynamically add dependencies and avoid it by dynamically installing packages on start but it makes completely no sense when you know how containers work and what bad side effects it can have (including security and performance problem) so we very strongly discourage this approach. Building your own image is the only reasonable way.
